core/src/main/java/io/questdb/cairo/map/SpillableHeapMemory.java

```diff
@@ -0,0 +1,402 @@
+    }
+
+    public void putNullBin32(long offset) {
+        putInt(offset, TableUtils.NULL_LEN);
+    }
+
+    public void putStrLowerCase(long offset, CharSequence value) {
+        if (value != null) {
+            putStrLowerCase(offset, value, 0, value.length());
+        } else {
+            putNullStr(offset);
+        }
+    }
+
+    public void putStrLowerCase(long offset, CharSequence value, int pos, int len) {
+        final long addr = appendAddressFor(offset, Vm.getStorageLength(len));
+        Unsafe.getUnsafe().putInt(addr, len);
+        Chars.copyStrCharsLowerCase(value, pos, len, addr + Vm.STRING_LENGTH_BYTES);
+    }
+
+    @Override
+    public long resize(long size) {
+        extend(size);
+        return mem.getAddress();
+    }
+
+    @Override
+    public void shiftAddressRight(long shiftRightOffset) {
+        mem.shiftAddressRight(shiftRightOffset);
+    }
+
+    @Override
+    public long size() {
+        return mem.size();
+    }
+
+    /**
+     * Skips given number of bytes. Same as logically appending 0-bytes. Advantage of this method is that no memory
+     * write takes place.
+     *
+     * @param bytes number of bytes to skip
+     */
+    @Override
+    public void skip(long bytes) {
+        checkAndExtend(getAppendAddress() + bytes);
+        mem.skip(bytes);
+    }
+
+    /**
+     * Reset the appendOffset and resize to 1 page
+     */
+    @Override
+    public void truncate() {
+        if (isSpilled) {
+            mem.close();
+            ff.remove(spillPath);
+            mem = initMem();
+            isSpilled = false;
+        }
+        mem.truncate();
+        assert mem.size() == pageSize;
+    }
+
+    @Override
+    public void zero() {
+        mem.zero();
+    }
+
+    private long candidateExtendedSpilledSize(long size) {
+        final long nPages = size > 0 ? ((size - 1) >>> extendSegmentMsb) + 1 : 1;
+        size = nPages << extendSegmentMsb;
+        if (size > maxSize) {
+            throw LimitOverflowException.instance().put("Maximum memory sizes (").put(maxSize).put(
+                    ") breached in VirtualMemory");
+        }
+        return size;
+    }
+
+    private void checkAndExtend(long address) {
+        if (address <= mem.getAddress() + mem.size()) {
+            return;
+        }
+        long sz = address - mem.getAddress();
+        if (!isSpilled && (sz > diskSpillThreshold)) {
+            final MemoryCARW cmarw = initFileBasedMem(candidateExtendedSpilledSize(sz), mem.getAppendOffset());
+            Vect.memmove(cmarw.getAddress(), mem.getAddress(), mem.size());
+            LOG.debug().$("Start spilling. curSz: ").$(mem.size()).$(", requestedSz: ").$(sz).$(", newSz: ").$(
+                    cmarw.size()).$();
+            mem.close(); // close CARW
+            mem = cmarw;
+            isSpilled = true;
+            return;
+        }
+
+        // MemoryCMARWImpl does not have memory breached checking, so adding one here. In addition, default
```

Comment:
Hi Andrei,
Sorry for late follow up.
Id like to clarify more detail about the extend behavior since Ive been overthinking about this. To start, this is my current understanding:
- `pageSize` should be power of two
- in the in-mem mode, it grows as a power of two.
- `extendSegmentSize` should also be power of two.
- when it grows, it grows as a multiple of extendSegmentSize (thus the size of mmapped heap is not necessarily power of two)

Then, 
- In FastMap, the `MAX_HEAP_SIZE` is defined as `(Integer.toUnsignedLong(-1) - 1) << 3`. However, it is not a power of 2 (which might also not an exact multiples of `extendSegmentSize`). How do you think the extend behavior should be when extending up to `MAX_HEAP_SIZE`?
Is it okay to allow memory to extend beyond the MAX_HEAP_SIZE? And if it is allowed, is special handling required so that only part of the allocated memory can be used instead of all (in this case, we only use the memory up to MAX_HEAP_SIZE rather than its actual allocated size)

Reply:
In addition, lets say the requested size is n * pageSize, what should be the allocated size?
In current implementation, CARW and CMARW has different implementation.
- CARW: requested n*pageSize -> given n*pageSize
- CMARW: requested n*pageSize -> given (n+1)*pageSize

Reply:
> In addition, lets say the requested size is `n * pageSize`, what should be the allocated size?

In file-based mode, we should grow the file in pages (page size should be multiple of the OS page, see `Files.PAGE_SIZE`). In in-memory mode, we should keep existing logic with the power of 2. In both cases, the allocated size should be equal or larger than the requested size.

Reply:
Thanks for the clarification. So, lets say the `extendSegmentSize` is set to 4GB (default value you proposed earlier) and the current heap size is 28GB (7 pages). When map is required to grow to 29GB, for example, since it extends based on page size, it will never be allowed since the next target size is 32GB = 34,359,738,368, while the max allowed heap is `Integer.toUnsignedLong(-1) - 1) << 3 = 34,359,738,352`.
In that case, is it okay to just throw error, or should it be allowed to grow to the MAX_HEAP_SIZE?

Reply:
> In that case, is it okay to just throw error, or should it be allowed to grow to the MAX_HEAP_SIZE?

I'd say that if the requested size (29GB) is under `MAX_HEAP_SIZE`, we should grow to `MAX_HEAP_SIZE`. After that, the next growth attempt should fail.

> So, lets say the `extendSegmentSize` is set to 4GB (default value you proposed earlier)

I recall that I mentioned 4GB as the disk spill threshold, not the `extendSegmentSize`. I'd make `extendSegmentSize` something in the 256-512MB range by default. The idea is to do less costly file growth operations, while not growing it too rapidly to avoid using the disk space too eagerly.

Reply:
> I recall that I mentioned 4GB as the disk spill threshold, not the extendSegmentSize. I'd make extendSegmentSize something in the 256-512MB range by default. The idea is to do less costly file growth operations, while not growing it too rapidly to avoid using the disk space too eagerly.

I see

> I'd say that if the requested size (29GB) is under MAX_HEAP_SIZE, we should grow to MAX_HEAP_SIZE. After that, the next growth attempt should fail.

In summary, does that mean the last extension may extend without respecting the `extendSegmentSize`?
Or in other word, should we extend up to MAX_HEAP_SIZE, or up to the size below MAX_HEAP_SIZE which is multiple of `extendSegmentSize`?

Reply:
> In summary, does that mean the last extension may extend without respecting the `extendSegmentSize`?
> Or in other word, should we extend up to MAX_HEAP_SIZE, or up to the size below MAX_HEAP_SIZE which is multiple of `extendSegmentSize`?

Yes, that sounds logical to me. File extension should try to use as much heap space as possible.